### PR TITLE
Harden package put route

### PR DESCRIPTION
**Pull request by trygve-lie**
This hardens the package put route and the process of uploading, extracting and saving the content of a package. 

It does now work so that if validation is not passed, a file for upload is missing or the wrong format the route will be terminated with a http status error. If we get past that point a stream is returned and we will start writing to the stream for each file written to the sink. If something fail during extraction of the package and writing to the sink, the stream will be terminated.

Due to the inner workings of the module parsing the uploaded file, we need to do some tracking of where we are in the upload, extraction and persisting process. This is done by keeping a state on a `._state` property.

Its is also introduced a log which is persisted as a json along with the extracted files.

All cover cases is not covered and some nits and bits is missing, but its getting wonky writing the rest without writing tests. I would like to merge this before proceeding with the rest of the corner cases due to the fact that multiple tests are merged into master and I do not want master and this work to be too out of sync.

### CODE REVIEW

File: lib/handlers/pkg.put.js [hunk 1]
```diff
@@ -41,50 +65,57 @@ const Parser = class Parser extends Duplex {
...
-                this.emit('done');
-            });
-        });
 
-        this.parser.on('partsLimit', () => {
-            const error = new Error('Payload Too Large')
-            this.destroy(error);
-        });
+                // Uploaded file was never extracted, but tar module did not error.
+                // Iow; we did not get a tar file. Trigger error.
```
Comment (digitalsadhu) [inline lib/handlers/pkg.put.js]: great to have these comments in here 👍

File: lib/handlers/pkg.put.js [hunk 2]
```diff
@@ -93,54 +124,69 @@ const Parser = class Parser extends Duplex {
...
-            ':version is a required url parameter and must be a string'
-        );
+        const outgoing = new HttpOutgoing();
+        outgoing.mimeType = 'application/octet-stream';
+
+        pipeline(parser, outgoing, (error) => {
+            if (error && parser._state === STATE_INITIALIZING) {
+                reject(error);
+                return;
+            }
```
Comment (digitalsadhu) [inline lib/handlers/pkg.put.js]: what happens if error but state is not STATE_INITIALIZING?

File: lib/utils/utils.js [hunk 3]
```diff
@@ -31,7 +31,7 @@ const readJSON = (sink, path) => {
 module.exports.readJSON = readJSON;
 
 const writeJSON = (sink, path, obj, contentType) => {
-    return new Promise((resolve, reject) => {
+    return new Promise(async (resolve, reject) => {
```
Comment (digitalsadhu) [inline lib/utils/utils.js]: link will probably moan about using an async function in a new Promise call. Might have to ignore

File: lib/handlers/pkg.put.js (hunk 2 above)
Comment (trygve-lie) [reply lib/handlers/pkg.put.js]: Then one are passed the point where a http 4xx or 5xx can be sent. A http 200 is already sent and some data might have been pushed to the outgoing stream (ex; half a package is extracted and it then fails for some reason). What will happen then is that the stream is terminated. We are at a point where we can not send a error message to the user.

We should though log it server side.
